src/Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs

```diff
@@ -842,13 +842,25 @@ private async Task BuildAndReport()
                 {
                     // The build was likely cancelled. We do not need to log an error in this case.
                 }
-                else if (_projectLoggingContext is null)
+                else if (ex is InternalLoggerException)
                 {
-                    _nodeLoggingContext.LogError(BuildEventFileInfo.Empty, "UnhandledMSBuildError", ex.ToString());
+                    string realMessage = ex.Message;
+                    Exception realEx = ex;
+                    while (realEx.InnerException is not null)
+                    {
+                        realEx = realEx.InnerException;
+                        realMessage = realEx.Message;
+                    }
```

Comment:
Are we interrested in the innermost exception here? Or can we reuse https://github.com/dotnet/msbuild/blob/main/src/Shared/TaskLoggingHelper.cs#L511 ?

Reply:
I was originally thinking I could use its stack trace, too, but reusing code is very tempting...

I'm fine with using the outermost exception's stack trace and the message from TaskLoggingHelper or keeping it as-is. Do you have a preference?

Reply:
Option 3 is modifying TaskLoggingHelper.GetInnerExceptionMessageString to also return the innermost exception's call stack. Then I could use both from that?

Reply:
I personally prefer as much diagnostic info as possible - so we should have `Exception.ToString` somewhere.

Users can sometime got scared by seeing stack traces - that's the reason why the current helper unwraps the messages only and the full `ToString()` version is logged just as a low-pri message. So you can probably follow that pattern. Though - in this specific case - when someone explicitly attaches additional log that has issues - I personaly wouldn't mind even if the full `ToString()` is logged as error. Your choice then :-)

Reply:
@Forgind Just a ping to see if you are aware and plan to make changes.

Reply:
Thanks for the ping—I think I like reusing TaskLoggingHelper.GetInnerExceptionMessageString. I don't think the error should be super verbose, so I'm satisfied with leaving it there and logging a low-pri message afterwards with the full ToString.

Reply:
Thank you @Forgind for the update